samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java

```diff
@@ -105,96 +102,222 @@ public void init(SamzaContainerContext containerContext, TaskContext taskContext
+        } else if (!isWriteAround) {
+          if (value == null) {
+            cache.delete(key);
+          } else {
+            cache.put(key, value);
+          }
+        }
+        writeMetrics.putNs.update(System.nanoTime() - startNs);
+        return result;
+      });
+  }
+
+  @Override
+  public void putAll(List<Entry<K, V>> records) {
     try {
-      lock.lock();
-      rwTable.put(key, value);
-      if (!isWriteAround) {
-        cache.put(key, value);
-      }
-    } finally {
-      lock.unlock();
+      putAllAsync(records).get();
+    } catch (InterruptedException e) {
+      throw new SamzaException(e);
+    } catch (Exception e) {
+      throw (SamzaException) e.getCause();
     }
-    writeMetrics.putNs.update(System.nanoTime() - startNs);
   }
 
   @Override
-  public void putAll(List<Entry<K, V>> entries) {
+  public CompletableFuture<Void> putAllAsync(List<Entry<K, V>> records) {
     writeMetrics.numPutAlls.inc();
     long startNs = System.nanoTime();
     Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
-    entries.forEach(e -> put(e.getKey(), e.getValue()));
-    writeMetrics.putAllNs.update(System.nanoTime() - startNs);
+    return rwTable.putAllAsync(records).handle((result, e) -> {
+        if (e != null) {
+          throw new SamzaException("Failed to put records " + records, e);
+        } else if (!isWriteAround) {
+          Set<K> allKeys = new HashSet<>();
+          List<K> delKeys = new ArrayList<>();
+          List<Entry<K, V>> putRecords = new ArrayList<>();
+          // Separate out put vs delete records
+          records.forEach(r -> {
+              allKeys.add(r.getKey());
+              if (r.getValue() != null) {
+                putRecords.add(r);
+              } else {
+                delKeys.add(r.getKey());
+              }
+            });
+
+          cache.deleteAll(delKeys);
+          cache.putAll(putRecords);
+        }
+
+        writeMetrics.putAllNs.update(System.nanoTime() - startNs);
+        return result;
+      });
   }
 
   @Override
   public void delete(K key) {
+    try {
+      deleteAsync(key).get();
+    } catch (InterruptedException e) {
+      throw new SamzaException(e);
+    } catch (Exception e) {
+      throw (SamzaException) e.getCause();
+    }
+  }
+
+  @Override
+  public CompletableFuture<Void> deleteAsync(K key) {
     writeMetrics.numDeletes.inc();
     long startNs = System.nanoTime();
     Preconditions.checkNotNull(rwTable, "Cannot delete from a read-only table: " + rdTable);
-    Lock lock = stripedLocks.get(key);
+    return rwTable.deleteAsync(key).handle((result, e) -> {
+        if (e != null) {
+          throw new SamzaException("Failed to delete the record for " + key, e);
+        } else if (!isWriteAround) {
+          cache.delete(key);
+          writeMetrics.deleteNs.update(System.nanoTime() - startNs);
```

Comment:
looks like we're only updating the timer metric if it's a writeAround cache. I'm not sure if we should do that - metrics should be updated regardless since you're measuring the time it took for the "delete" operation right?